### ISSUE TITLE
don't allow arbitrary sprixel placement within a plane

### ISSUE THREAD

**Issue by dankamongmen** [dankamongmen/notcurses#1572]
we currently accept `y` and `x` in `ncvisual_options` for `NCBLIT_PIXEL` with a non-`NULL` `vopts->n`. i don't think this is a good idea (and it's currently broken if the underlying plane gets resized, because what do planes know about these variables?). the `y` and `x` fields in `sprixel` are bastards, poorly understood and almost always specified as 0. they ought be purged. reject an attempt to place a sprixel within a preexisting plane at any location but the origin. support `y` and `x` for the `NULL` case of `vopts->n`, as they specify where to place the resulting plane.

the only problem is that `xray` makes use of this, blitting to the standard plane (and i thought we didn't allow sprixels to blit to the standard plane, either?). so that needs be changed.

**Comment by joseluis** [comment]
Are you sure you don't want to rethink #1109 for 3.0 ? It somehow reminds me of the changes Python went through from 2 to 3 too.

**Comment by dankamongmen** [comment]
> Are you sure you don't want to rethink #1109 for 3.0 ? It somehow reminds me of the changes Python went through from 2 to 3 too.

man what have you got against the standard plane? =]

i've thought about it a few times, actually, and i just don't see any win *except* conceptual simplicity. conceptual simplicity is great, but not when it is a real hit to usability (IMHO, and for this purpose). without a standard plane, here's what a new user has to do just to get some text up:

* learn how to create a plane. ok, let's look at `ncplane_create()`. wait, that already requires a `ncplane`, and i don't have one.
* look at `ncpile_create()`. ok, so what's a pile? oh, this function doesn't return a pile? it returns a plane? but a pile now exists? ok
* oh shit i resized my screen but it didn't resize my plane. how do i deal with that?
* a resize callback? ok so i need create this with a resize callback that fits me to the terminal.
* how do i make this show up? oh ok i call `notcurses_render()`
* oh wait `notcurses_render()` doesn't exist in this standard planeless world
* gotta call `ncpile_render()` then, ok, these piles again...wait i don't have a pile to feed it...wait it accepts `ncplane`s lol ok

i mean it's not so bad, but *almost every program wants a standard plane with the standard plane semantics*. and if you don't want it, you just ignore it. send it to the bottom, or work on a fresh pile from the beginning.

it's only coming up here as an instance of the general problem of "can sprixels be rendered to preexisting planes?" which i have to allow to allow arbitrary sizing (now, if we added two fields to `ncvisual_options`, we could get away without that, and require new planes every time. but that's still not great, because sometimes you want to do different sprixels, but must retain the TAM to avoid flicker.) but rendering to arbitrary positions in preexisting planes is another thing entirely, whether the standard plane or some other one.

i updated the bug to include the phrase "at any location [but the origin]", which i think makes this clearer.

basically, every time i ask about getting rid of the standard plane, i flip it around and ask, "if we did not have a concept of a standard plane, would i add it?" and the answer is always yes.

**Comment by dankamongmen** [comment]
so really this doesn't seem all that unreasonable. if `ncplane_resize()` is broken, it ought be fixed. we ought go through and check everywhere that we change coordinate systems, and verify that they're using this correctly. but i see no fundamental problem with this.

**Comment by dankamongmen** [comment]
better yet would be to put some solid unit testing on it!

**Comment by dankamongmen** [comment]
hrmm we'd need know where the preserved TAM is within the `ncplane`, since the new blit could be at a different place, in which case the TAM wouldn't wholly apply (and would definitely need translation). hrmm. maybe not a great idea after all.

**Comment by dankamongmen** [comment]
wouldn't the ability to do this kinda imply the ability to put more than one sprixel on a plane? which we are definitely not ever supporting.

**Comment by dankamongmen** [comment]
if we decide to lock it down, restructure `paint()` to do the sprixel check once, and then just run that methodology. right now we check for sprixel once per cell on such a plane, and then run `sprixel_by_id()` once per cell as well. we could fold that all down into O(1).

**Comment by dankamongmen** [comment]
i think maybe the restriction ought be: you can bring your own plane, and it can even be the standard plane, but you can only do that if your render exactly fills the provided plane (modulo sixel jankiness). that means either custom-fitting it, or using `NCSCALE_STRETCH`. that eliminates the standard plane special case, eliminates just about everything ugly about bring-your-own-plane, solves this problem for us (nope, we only allow sprixels at origins), and eliminates just about every other open question.

yes. i think this is the way.

**Comment by joseluis** [comment]
I think I like that, as long as it's clearly documented.

BTW what does *TAM* mean?

**Comment by dankamongmen** [comment]
> I think I like that, as long as it's clearly documented.

yeah, this is all going to get a writeup in `notcurses_plane(3)` and probably `notcurses(3)`, definitely `notcurses_visual(3)`. let me file a bug on doing that.

> BTW what does _TAM_ mean?

It is a harbinger of dashed hopes, and idle dreams.

> It may be that universal history is the history of the different intonations given a handful of metaphors.
― Jorge Luis Borges

https://nick-black.com/dankwiki/index.php?title=Theory_and_Practice_of_Sprixels#The_T-A_Matrix
https://github.com/dankamongmen/notcurses/blob/master/src/lib/internal.h#L48 through L142

**Comment by dankamongmen** [comment]
if we're going this way, make sure to kill off `tacachey` and `tacachex`, which become redundant.

**Comment by joseluis** [comment]
> https://github.com/dankamongmen/notcurses/blob/master/src/lib/internal.h#L48 through L142

Just in case I blow your mind https://github.com/dankamongmen/notcurses/blob/master/src/lib/internal.h#L48-L142

**Comment by dankamongmen** [comment]
> Just in case I blow your mind https://github.com/dankamongmen/notcurses/blob/master/src/lib/internal.h#L48-L142

https://www.youtube.com/watch?v=QgmzEhXe7J4

**Comment by dankamongmen** [comment]
how's this

```
Some terminals support pixel-based output via one of a number of protocols.                   
Checking for bitmap support requires interrogating the terminal and reading a                 
response. This takes time, and will never complete if the terminal doesn't                    
respond. Notcurses will not do so without an explicit request from the client                 
code. Before **NCBLIT_PIXEL** can be used, it is thus necessary to call                       
**notcurses_check_pixel_support**. If this function has not successfully                      
returned, attempts to use **NCBLIT_PIXEL** will fall back to cell-based                       
blitting (or fail, if **NCVISUAL_OPTION_NODEGRADE** is used).                                 
                                                                                              
**NCBLIT_PIXEL** has some stringent requirements on the type of planes it can                 
be used with; it is usually best to let **ncvisual_render** create the backing                
plane by providing a **NULL** value for **n**. If you must bring your own                     
plane, that plane must not have a resize callback, and must be perfectly sized                
for the bitmap (i.e. large enough, and not more than a full cell larger in                    
either dimension). Using **NCSCALE_STRETCH** means that the second condition                  
will always be met. Once a sprixel is blitted to a plane, that plane may not                  
be resized, a resize callback may not be set, the base cell may not be set,                   
and cell methods (including cell blitting) may not be used with it. The only                  
way to change the plane's contents is to destroy the plane. Notcurses will                    
automatically set the plane's base cell to the null glyph, **NCSTYLE_NONE**,                  
and transparent in both the fore- and background.                                             
                                                                                              
Only one bitmap can be blitted onto a plane at a time (but multiple planes                    
with bitmaps may be visible); blitting a second to the same plane will delete                 
the original. Destroying the plane with which a bitmap is associated will                     
delete the bitmap.
```

**Comment by dankamongmen** [comment]
So, the final word on this is:

* no resizing of sprixelated planes
* if providing your own plane for sprixelation, it must be large enough
* if providing your own plane for sprixelation, it must not be too large (every cell must be at least partially covered by the sprixel)
* no cell-based output to sprixelated planes

so let it be documented, so let it be enforced, ideally by a typing system but runtime conditionals will do in a pinch

**Comment by dankamongmen** [comment]
hrmmm, we probably want to allow resizing of a sprixelated plane so long as the sprixel is likewise resized. no....we don't have the `ncvisual` source material, let alone the original media, so we couldn't do a proper resize anyway in many cases. hrmmm, that's going to have ramifications for e.g. `ncplayer -bpixel`, which used to stretch to fit the viewing area (it now stays the same size, whether the visible area shrinks or grows). hrmmmph.

when we normally resize in `ncplayer`, it's not like we replay the *current* frame; it takes effect on the *next* frame...but i don't see how that can help.

can we get by with just a restriction that the sprixel be at the origin? but the plane can be bigger? that eliminates the need to keep an offset for the TAM within a plane, or for the sprixel...but then we do need keep the geometry of the TAM. unless we always enforce the *TAM* being the same size as the plane.

come to think of it, if we're going to allow pixel bltiting to the standard plane, then we *can't* enforce a "no resize" rule, because the standard plane **must** be resized when the visual area is resized. so that decides for us: resizing is allowed. perhaps add a new `sprixcell_e`, `SPRIXCELL_ABSENT`. so what happens on a resize? we can't do anything useful on a grow, and on a shrink, the only question is how much to cut. i say all of it -- on a resize, a plane is desprixelated, and reverts to a real human boy. this will apply to either a shrink or a grow. yes, i think *this* is the Way.

**Comment by dankamongmen** [comment]
There are probably numerous other glyph entry points we need shore up for sprixelated planes, but we've got `ncplane_putc_yx()` handled and unit tested. All `ncplane_put*()` functions go through there, so it's good initial coverage.
